File: backend/src/main/java/app/miniappspring/controller/CategoryController.java
package app.miniappspring.controller;

import app.miniappspring.dto.product.ProductCardDto;
import app.miniappspring.dto.product.category.CategoryDto;
import app.miniappspring.dto.product.category.NumberOfProductsInThisCategory;
import app.miniappspring.service.CategoryService;
import app.miniappspring.service.ProductService;
import jakarta.servlet.http.HttpServletRequest;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/v1/category")
@RequiredArgsConstructor
public class CategoryController {

    private final CategoryService categoryService;
    private final ProductService productService;

    @GetMapping("/categories")
    public List<String> getAllCategories() {
        return categoryService.getAllCategories();
    }

    @GetMapping
    public NumberOfProductsInThisCategory getMapKeyCategoryValueCountProduct() {
        return categoryService.getCategories();
    }

    @PreAuthorize("hasAnyAuthority('ROLE_EMPLOYEE')")
    @PostMapping("/add")
    public ResponseEntity<String> createCategoryProduct(@RequestParam String categoryName, HttpServletRequest request) {
        return ResponseEntity.ok().body(categoryService.createCategoryProduct(categoryName));
    }

    @PostMapping("/products")
    public List<ProductCardDto> getProductsByCategory(@RequestBody CategoryDto category) {
        return productService.getProductsByCategory(category);
    }
}

File: frontend/react-app/src/controller/CategoryController.ts
import api from "../http";
import {CategoryResponse} from "../model/response/product/CategoryResponse";

export class CategoryController {

    public static async getCategories(){
        return await api.get<Array<string>>("/api/v1/category/categories").then(r => r.data);
    }

    static async getMapCategory(){
        return api.get<CategoryResponse>('/api/v1/category')
            .then(response=>response.data);
    }


    public static async createCategory(categoryName:string):Promise<string>{
        return api.post<string>('/api/v1/category/add?categoryName='+categoryName).then(r=>r.data);
    }
}